### PR TITLE
Fix rename

### PR DESCRIPTION
**Pull request by Ruadhri17**
## Proposed changes
Replace rename with copy and delete where necessary. Calculate tmp path inside atomically write functions. More in #1899

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#1899
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/common/tedge_utils/src/file.rs
Comment (didier-wenzek) [inline crates/common/tedge_utils/src/file.rs:118]: I would not directly copy/delete the file but try first to rename it as the copy/delete will not work if the file is larger than the free space (this is not unlikely with a firmware download on a resource-constraint device).

If would be better to fallback to a copy/delete only if the rename fails because of a [cross-filesystem rename](https://doc.rust-lang.org/nightly/std/io/enum.ErrorKind.html#variant.CrossesDevices), but unfortunately this is a nightly feature. So just try to rename and copy/delete in case of some error.

**Review by didier-wenzek — COMMENTED** [review]
That's perfect. However, for the move_file function it's better to try first a simple rename and use a copy/delete only if the filesystems are different.

File: crates/common/download/src/download.rs
Comment (rina23q) [reply crates/common/download/src/download.rs:84]: Hmm in `c8y-configuration-plugin`, there is a parameter `perm`, where user can decide which permission they want to have in a downloaded file. I feel this change is breaking it.

**Review by rina23q — COMMENTED** [review]
I have a doubt if `target_permission` can really removed from `Downloader`. I will run a test manually and come back.

**Review by rina23q — CHANGES_REQUESTED** [review]
Tested and unfortunately my concern was correct. I would block merging until the regression fixed.

**Comment by didier-wenzek** [comment]
> Tested and unfortunately my concern was correct. I would block merging until the regression fixed.

As these permission settings are used in a single place, the simpler it's to keep the signature of the  `move_file` as proposed by @Ruadhri17 and to let the configuration manager [applies the requested permission](https://github.com/thin-edge/thin-edge.io/blob/main/crates/common/tedge_utils/src/file.rs#L150) on the file once moved.

File: crates/common/tedge_utils/src/file.rs
Comment (Ruadhri17) [reply crates/common/tedge_utils/src/file.rs]: Changed.

**Comment by didier-wenzek** [comment]
> Oops, I just reverted changes regarding `PermissionEntry` smile, though I can bring them back. If we decide to stick with my solution, should we solve the problem with `c8y-configuration-plugin` in this PR or create the new one?

This must be fixed in that PR. But don't revert your revert: let @rina23q have a look first. I will do the same.

File: crates/common/tedge_utils/src/file.rs
Comment (didier-wenzek) [inline crates/common/tedge_utils/src/file.rs:99]:
```suggestion
///     Ok() when file was moved successfully
```

File: crates/common/tedge_utils/src/file.rs
Comment (didier-wenzek) [inline crates/common/tedge_utils/src/file.rs:143]: This must also be done in the copy-and-delete case.

**Review by didier-wenzek — COMMENTED** [review]
You can keep unchanged the type signature of the `move_file` i.e to keep the `new_file_permissions: PermissionEntry` parameter. However setting these permissions must be done independently of the actual method used: rename or copy-and-delete.

File: crates/common/tedge_utils/src/file.rs
Comment (rina23q) [reply crates/common/tedge_utils/src/file.rs:143]: Agree with @didier-wenzek

**Review by rina23q — COMMENTED** [review]
The regression is fixed by the reverting `move_file` change, however, some still need to be changed before merge.

**Comment by rina23q** [comment]
Note: if we had https://github.com/thin-edge/thin-edge.io/pull/1875, the permission issue would have been tested by pipeline...

**Review by didier-wenzek — APPROVED** [review]
Approved.

**Pull request merged**
